SetMultipleResponseHeaders still creates a clone constraint on the response body

- [x] I have looked for existing issues (including closed) about this

## Bug Report

#703 's description and the changelog both say that this was fixed for Request and Response types, but the PR only actually applied the fix to the request types. For me this means that in axum I still can't use `SetMultipleResponseHeadersLayer` because axum's `Router:: layer` requires `L: Layer<Route> + Clone + Sync + 'static`.

### Version

0.7.0, main

### Platform

Darwin mac.lan 25.5.0 Darwin Kernel Version 25.5.0: Mon Apr 27 20:41:15 PDT 2026; root:xnu-12377.121.6~2/RELEASE_ARM64_T6041 arm64 arm Darwin

### Description

Here's a reproduction:

```rust
use axum::{Router, routing::get, http::{header, HeaderValue}};
use tower_http::set_header::response::SetMultipleResponseHeadersLayer;

fn build() -> Router {
    Router::new()
        .route("/", get(|| async { "ok" }))
        .layer(SetMultipleResponseHeadersLayer::overriding(vec![
            (header::X_CONTENT_TYPE_OPTIONS, HeaderValue::from_static("nosniff")).into(),
        ]))
}
```
```console
error[E0277]: the trait bound `Response<Body>: Clone` is not satisfied
   --> src/main.rs:7:16
    |
  7 |           .layer(SetMultipleResponseHeadersLayer::overriding(vec![
    |  __________-----_^
    | |          |
    | |          required by a bound introduced by this call
  8 | |             (header::X_CONTENT_TYPE_OPTIONS, HeaderValue::from_static("nosniff")).into(),
  9 | |         ]))
    | |__________^ the trait `Clone` is not implemented for `Response<Body>`
    |
    = note: required for `Response<Body>` to implement `Clone`
    = note: 1 redundant requirement hidden
    = note: required for `SetMultipleResponseHeadersLayer<Response<Body>>` to implement `Clone`
note: required by a bound in `Router::<S>::layer`
   --> /Users/mumu/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/axum-0.8.9/src/routing/mod.rs:305:27
    |
303 |     pub fn layer<L>(self, layer: L) -> Router<S>
    |            ----- required by a bound in this associated function
304 |     where
305 |         L: Layer<Route> + Clone + Send + Sync + 'static,
    |                           ^^^^^ required by this bound in `Router::<S>::layer`
help: consider borrowing here
    |
  7 |         .layer(&SetMultipleResponseHeadersLayer::overriding(vec![
    |                +
```

Happy to review a change to add it to the response type as well.

On Wed, Jul 15, 2026, 12:07 PM Murph Murphy ***@***.***>
wrote:

> *skeet70* created an issue (tower-rs/tower-http#713)
> <https://github.com/tower-rs/tower-http/issues/713>
>
>    - I have looked for existing issues (including closed) about this
>
> Bug Report
>
> #703 <https://github.com/tower-rs/tower-http/pull/703> 's description and
> the changelog both say that this was fixed for Request and Response types,
> but the PR only actually applied the fix to the request types. For me this
> means that in axum I still can't use SetMultipleResponseHeadersLayer
> because axum's Router:: layer requires L: Layer<Route> + Clone + Sync +
> 'static.
> Version
>
> 0.7.0, main
> Platform
>
> Darwin mac.lan 25.5.0 Darwin Kernel Version 25.5.0: Mon Apr 27 20:41:15
> PDT 2026; root:xnu-12377.121.6~2/RELEASE_ARM64_T6041 arm64 arm Darwin
> Description
>
> Here's a reproduction:
>
> use axum::{Router, routing::get, http::{header, HeaderValue}};use tower_http::set_header::response::SetMultipleResponseHeadersLayer;
> fn build() -> Router {
>     Router::new()
>         .route("/", get(|| async { "ok" }))
>         .layer(SetMultipleResponseHeadersLayer::overriding(vec![
>             (header::X_CONTENT_TYPE_OPTIONS, HeaderValue::from_static("nosniff")).into(),
>         ]))}
>
> error[E0277]: the trait bound `Response<Body>: Clone` is not satisfied   --> src/main.rs:7:16    |  7 |           .layer(SetMultipleResponseHeadersLayer::overriding(vec![    |  __________-----_^    | |          |    | |          required by a bound introduced by this call  8 | |             (header::X_CONTENT_TYPE_OPTIONS, HeaderValue::from_static("nosniff")).into(),  9 | |         ]))    | |__________^ the trait `Clone` is not implemented for `Response<Body>`    |    = note: required for `Response<Body>` to implement `Clone`    = note: 1 redundant requirement hidden    = note: required for `SetMultipleResponseHeadersLayer<Response<Body>>` to implement `Clone`note: required by a bound in `Router::<S>::layer`   --> /Users/mumu/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/axum-0.8.9/src/routing/mod.rs:305:27    |303 |     pub fn layer<L>(self, layer: L) -> Router<S>    |            ----- required by a bound in this associated function304 |     where305 |         L: Layer<Route> + Clone + Send + Sync + 'static,    |                           ^^^^^ required by this bound in `Router::<S>::layer`help: consider borrowing here    |  7 |         .layer(&SetMultipleResponseHeadersLayer::overriding(vec![    |                +
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/tower-rs/tower-http/issues/713?email_source=notifications&email_token=AKWOC6WQXKMQJEN4IOOVDUD5E7JAFA5CNFSL4Z3JMQ5C6L3HNF2C22DVMIXUS43TOVSS6NBYHE2TKNJRG43DTJTSMVQXG33OVJZXKYTTMNZGSYTFMSSWK5TFNZ2KYZTPN52GK4S7MNWGSY3L>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AKWOC6QPVSYQ54DLQFVBELD5E7JAFAVCNFSNUABFKJSXA33TNF2G64TZHMYTCOBVGE3TQMJTHNEXG43VMU5TIOBZGU2TKMJXGY42C5QC>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>

--- a/tower-http/src/set_header/response/multiple_headers.rs
+++ b/tower-http/src/set_header/response/multiple_headers.rs
@@ -18,9 +18,16 @@ use crate::set_header::{HeaderInsertionConfig, HeaderMetadata, InsertHeaderMode}
 /// Layer that applies [`SetMultipleResponseHeader`] which adds multiple response headers.
 ///
 /// See [`SetMultipleResponseHeader`] for more details.
-#[derive(Clone)]
 pub struct SetMultipleResponseHeadersLayer<M> {
     headers: Vec<HeaderInsertionConfig<M>>,
+}
+
+impl<M> Clone for SetMultipleResponseHeadersLayer<M> {
+    fn clone(&self) -> Self {
+        Self {
+            headers: self.headers.clone(),
+        }
+    }
 }
 
 impl<M> fmt::Debug for SetMultipleResponseHeadersLayer<M> {
@@ -86,11 +93,21 @@ impl<S, M> Layer<S> for SetMultipleResponseHeadersLayer<M> {
 }
 
 /// Middleware that sets multiple headers on the response.
-
-#[derive(Clone)]
 pub struct SetMultipleResponseHeader<S, M> {
     inner: S,
     headers: Vec<HeaderInsertionConfig<M>>,
+}
+
+impl<S, M> Clone for SetMultipleResponseHeader<S, M>
+where
+    S: Clone,
+{
+    fn clone(&self) -> Self {
+        Self {
+            inner: self.inner.clone(),
+            headers: self.headers.clone(),
+        }
+    }
 }
 
 impl<S, M> SetMultipleResponseHeader<S, M> {
@@ -419,5 +436,19 @@ mod tests {
         let layer = SetMultipleResponseHeadersLayer::overriding(vec![meta]);
         let debug_str = format!("{:?}", layer);
         assert!(debug_str.contains("SetMultipleResponseHeadersLayer"));
+    }
+
+    #[test]
+    fn test_service_clone() {
+        struct NonCloneBody;
+        let svc = tower::ServiceBuilder::new()
+            .layer(SetMultipleResponseHeadersLayer::<Response<NonCloneBody>>::overriding(vec![]))
+            .check_clone()
+            .service(service_fn(|_: Request<NonCloneBody>| async move {
+                Ok::<_, Infallible>(Response::new(NonCloneBody))
+            }));
+
+        fn check_service_and_clone<T: Service<Request<NonCloneBody>> + Clone>(_: T) {}
+        check_service_and_clone(svc);
     }
 }

